source/slang/slang.cpp

```diff
@@ -6203,6 +6204,21 @@ void const* EndToEndCompileRequest::getEntryPointCode(int entryPointIndex, size_
     return (void*)blob->getBufferPointer();
 }
 
+SlangResult EndToEndCompileRequest::getCompileTimeProfile(ISlangBlob** compileTimeProfile)
+{
+    if (compileTimeProfile == nullptr)
+    {
+        return SLANG_E_INVALID_ARG;
+    }
+
+    StringBuilder outString;
+    SerialPerformaceProfiler::getProfiler()->accumulateResults(outString, PerformanceProfiler::getProfiler());
```

Comment:
Why do we need to accumulate in a mutex here? Can we simply return what is in PerformanceProfiler::getProfiler() ?

Reply:
Then how can the user get the serialized timing? They need to accumulate themselves?

Reply:
Well, they should because only they know how many threads are created. I actually did not understand when/how things are aggregated give the current implementation.

Reply:
Given current implementation, each slang working thread has to call this API. 
And then the main thread will get the result by calling this API again.

This is kind of like the parallel reduction.

Reply:
we don't need to return the string alternatively, but return the profiler instead. So user can just use the profiler data after `join()` for all the threads.

Basically, the usage for now is
```
void ThreadFunc()
{
     //... some slang work
     SerialPerformaceProfiler profiler;
     spGetCompileTimeProfile(&profiler);
}

void main()
{
      join(thread1);
      ....
      join(threadN);

      SerialPerformaceProfiler profiler;
      spGetCompileTimeProfile(&profiler);
       // code to print profiler
}
```

Reply:
I think what you're thinking about is that each thread will hold a thread local variable, and then the main thread will accumulate them together.
```
SerialPerformaceProfiler globalProfiler;
void ThreadFunc()
{
     //... some slang work
     SerialPerformaceProfiler profiler;
     spGetCompileTimeProfile(&profiler);

     {
         // scope lock
         // accumulate(globalProfiler, profiler);
     }
}
```

It just depends on whether we want to do the accumulation or user.

Reply:
@csyonghe can you suggest that which way we want?
Let slang do the accumulation or user do the accumulation?
IMO, either way is fine, we just need to document the behavior carefully. It's just trivial work to change it in the implementation.

Reply:
I also another thought to make the profile in a hierarchical way. That might be a follow-up change.

Reply:
If we decide to only measure the time per thread, then we don't need to use lock at all. So the ball will be passed to user.

Reply:
But if that is our decision, I don't think I will need this `SerialPerformaceProfiler` class at all.
But I cannot return a string anymore. Instead, I have to return the profiler to user, so that they can use the profiler to do the follow-up operation. But that will requires us to expose the an interface to user. So I might need to create a struct/class exposed to slang.h used to carry the result.

Reply:
I think we should let the user aggregate. We can still return a string by serializing the result into a json. But it may be easier to just expose another com interface to represent the structural profiling result.